File: pyforms_web/controls/control_workflow.py
import datetime
from pyforms_web.controls.control_base import ControlBase
import simplejson

class ControlWorkflow(ControlBase):

	def init_form(self): return "new ControlWorkflow('{0}', {1})".format( self._name, simplejson.dumps(self.serialize()) )

	@property
	def selected_operator(self):
		if not hasattr(self, '_selected_operator'): return None
		return self._selected_operator
	@selected_operator.setter
	def selected_operator(self, value):
		if not hasattr(self, '_selected_operator'): return
		if self._selected_operator!=value: self.mark_to_update_client()
		
		self._selected_operator = value

	@property
	def selected_link(self):
		if not hasattr(self, '_selected_link'): return None
		return self._selected_link
	@selected_link.setter
	def selected_link(self, value):
		if not hasattr(self, '_selected_link'): return
		if self._selected_link!=value: self.mark_to_update_client()
		
		self._selected_link = value



	@property
	def operator_selected_evt(self):
		return self._operator_selected_evt

	@operator_selected_evt.setter
	def operator_selected_evt(self, value):
		self._operator_selected_evt = value

	@property
	def operator_unselected_evt(self):
		return self._operator_unselected_evt

	@operator_unselected_evt.setter
	def operator_unselected_evt(self, value):
		self._operator_unselected_evt = value



	@property
	def link_selected_evt(self): return self._link_selected_evt
	@link_selected_evt.setter
	def link_selected_evt(self, value): self._link_selected_evt = value

	@property
	def link_unselected_evt(self): return self._link_unselected_evt
	@link_unselected_evt.setter
	def link_unselected_evt(self, value): self._link_unselected_evt = value




	def deleteSelected(self): self._delete_selected = True



	def serialize(self):
		res = super(ControlWorkflow, self).serialize()

		if self.selected_operator: 	res.update({'selected_operator': self.selected_operator})
		if self.selected_link: 		res.update({'selected_link': self.selected_link})
		if hasattr(self, '_delete_selected'):
			res.update({'deleteSelected': self._delete_selected})
			del self._delete_selected
		
		res.update({'operator_selected_evt': 1 if hasattr(self, 'operator_selected_evt') else 0})
		res.update({'operator_unselected_evt': 1 if hasattr(self, 'operator_unselected_evt') else 0})
		
		res.update({'link_selected_evt': 1 if hasattr(self, 'link_selected_evt') else 0})
		res.update({'link_unselected_evt': 1 if hasattr(self, 'link_unselected_evt') else 0})
		
		return res

	def deserialize(self, properties):
		super(ControlWorkflow, self).deserialize(properties)
		if 'selected_operator' in properties: 	self._selected_operator = properties['selected_operator']
		if 'selected_link' in properties: 		self._selected_link = properties['selected_link']


	

File: pyforms_web/web/static/pyformsjs/ControlWorkflow.js
class ControlWorkflow extends ControlBase{

	////////////////////////////////////////////////////////////////////////////////

	init_control(){
		var html = "<div id='"+this.place_id()+"' class='field control ControlWorkflow ui segment' ><label>"+this.properties.label+"</label><div id='"+this.control_id()+"' ></div></div>";
		this.jquery_place().replaceWith(html);

		var self = this;
		this.jquery().change(function(){
			self.basewidget.fire_event( self.name, 'update_control_event', false );
		});

		

		$( '#'+this.control_id() ).flowchart({ data: this.properties.value, multipleLinksOnOutput:true });

		if(this.properties.operator_selected_evt)
			$( '#'+this.control_id() ).flowchart({ onOperatorSelect: function(operatorId){
				$( '#'+self.control_id() ).flowchart('unselectLink');
				self.properties.selected_operator = operatorId;
					
				if( !self.properties.stop_operator_select_evt ){
					self.properties.no_selected_operator_update = 1;
					self.properties.no_selected_link_update 	= 1;

					self.properties.selected_operator = operatorId;
					self.properties.selected_link = undefined;
					self.basewidget.fire_event( self.name, 'operator_selected_evt', false );
					self.properties.no_selected_operator_update = undefined;
				}
				
				return true;
			} });

		if(this.properties.operator_unselected_evt)
			$( '#'+this.control_id() ).flowchart({ onOperatorUnselect: function(){
				self.properties.selected_operator = undefined;
					
				if( !self.properties.stop_operator_unselect_evt ){

					self.properties.no_selected_operator_update = 1;
					self.properties.no_selected_link_update 	= 1;
					
					self.basewidget.fire_event( self.name, 'operator_unselected_evt', false );
					self.properties.no_selected_operator_update = undefined;
				}
				return true;
			} });





		if(this.properties.link_selected_evt)
			$( '#'+this.control_id() ).flowchart({ onLinkSelect: function(linkId){
				

				$( '#'+self.control_id() ).flowchart('unselectOperator');
				self.properties.selected_link = linkId;
					
				if( !self.properties.stop_link_select_evt ){
					self.properties.no_selected_operator_update = 1;
					self.properties.no_selected_link_update 	= 1;
					
					self.properties.selected_link = linkId;
					self.properties.selected_operator = undefined;
					
					self.basewidget.fire_event( self.name, 'link_selected_evt', false );
					self.properties.no_selected_link_update = undefined;
				}

				
				return true;
			} });

		if(this.properties.link_unselected_evt)
			$( '#'+this.control_id() ).flowchart({ onLinkUnselect: function(){
				self.properties.selected_link = undefined;
					
				if( !self.properties.stop_link_unselect_evt ){
					self.properties.no_selected_operator_update = 1;
					self.properties.no_selected_link_update 	= 1;
					
					self.basewidget.fire_event( self.name, 'link_unselected_evt', false );
					self.properties.no_selected_link_update = undefined;
				}
				return true;
			} });
		
	};

	////////////////////////////////////////////////////////////////////////////////

	set_value(value){
		$( '#'+this.control_id() ).flowchart('setData', value);
	};

	////////////////////////////////////////////////////////////////////////////////

	get_value(){ 
		return $( '#'+this.control_id() ).flowchart('getData');
	};

	////////////////////////////////////////////////////////////////////////////////

	serialize(){
		this.properties.value = this.get_value();

		if( !this.properties.no_selected_operator_update ){
			var selected_operator = $( '#'+this.control_id() ).flowchart('getSelectedOperatorId');
			if( selected_operator ) this.properties.selected_operator = selected_operator;
		}

		if( !this.properties.no_selected_link_update ){
			var selected_link = $( '#'+this.control_id() ).flowchart('getSelectedLinkId');
			if( selected_link ) this.properties.selected_link = selected_link;
		}
		return this.properties; 
	};


	////////////////////////////////////////////////////////////////////////////////

	deserialize(data){
		$.extend(this.properties, data);


		this.set_value(this.properties.value);

		if( this.properties.selected_operator ){
			
			this.properties.stop_operator_select_evt = 1;
			$('#'+this.control_id() ).flowchart('selectOperator', this.properties.selected_operator);
			this.properties.stop_operator_select_evt = undefined;
		}


		if( this.properties.selected_link>=0){
			
			this.properties.stop_link_select_evt = 1;
			this.properties.stop_link_unselect_evt = 1;
			$('#'+this.control_id() ).flowchart('selectLink', this.properties.selected_link);
			this.properties.stop_link_select_evt = undefined;
			this.properties.stop_link_unselect_evt = undefined;
		}
		
		if( data.deleteSelected ){
			this.properties.stop_operator_unselect_evt = 1;
			this.properties.stop_link_unselect_evt = 1;
			$( '#'+this.control_id() ).flowchart('deleteSelected');
			this.properties.stop_operator_unselect_evt = undefined;
			this.properties.stop_link_unselect_evt = undefined;
		};
		
	};
	
}